### ISSUE TITLE
Libs created with ar.exe contain duplicated symbols after an isolated symbol is moved back to the unity

### ISSUE THREAD

**Issue by belkiss** [fastbuild/fastbuild#309]
Heya,

I found another issue with commit 1cc889e06bee91feba0fcf1bcedf997a931a1f7f.

ar.exe by default will add or update, but not remove symbols from the libs.
So if a lib contains a unity and and a cpp file isolated from it, after revert of that file, the lib will contain a duplicated symbol.

**Comment by ffulin** [comment]
ar.exe (unlike lib.exe) has some "interesting" behaviour which is that depending on the command line provided, it will just continue to add symbols rather than recreating the archive as you might expect. In that case, FASTBuild deleting the archive could cover up the bad command line.

So the possible options here are to have FASTBuild:
a) delete libs if using ar.exe (small performance cost)
b) warn if ar.exe's command line is missing the correct options
c) modify/add the correct cmd line option to ar.exe
d) Have FASTBuild do nothing, assuming that your cmd line args are correct

I'm leaning a little towards option d), simply because it's the least likely to have unintended side effect (if using an alternative ar.exe from another vendor for example). Having said that, it's not great that this is a bit of a gotcha.

Thoughts?

**Comment by belkiss** [comment]
Hello!

Could you indicate what the correct command line options are?
I'm currently calling it with "rcs", to replace/insert.

I've been looking into the documentation of ar, and cannot see how to have it recreate the archive with only what was passed to the command line.

Another option would be to call ar with "d" when an isolated file is moved back to the blob.

**Comment by ffulin** [comment]
I thought "rcs" was the correct cmd line, as FASTBuild for Linux and OSX is using that, but perhaps it has this problem as well.  I remember actually fixing this issues specifically by modifying the command line on a previous project, but I'd have to dig a bit more to find out what the exact args are. (I took a really quick look at the ar.exe docs and it didn't jump out at me).

**Comment by belkiss** [comment]
I can confirm that the linux version of FastBuild has the same problem.

**Comment by ffulin** [comment]
I spent some time trying to reproduce this under Linux, by opening for edit (bringing out of unity) files when compiling FASTBuild, and reverting them. I wasn't able to cause any problems with linking the executable (I made sure to change function signatures to avoid a false sense of security).

Do you have some good repro steps for this? All my cases seemed to link the final executables correctly - I recall when having this problem with the Orbis librarian (which missing the extra option I mentioned) I had duplicate symbol errors during linking - is that what you are seeing?

BTW - I'm now leaning towards restoring the deletion of the libs when using ar.exe, but I'd love to have a test case to cover that if possible, to be sure I've understood the problem.

**Comment by belkiss** [comment]
Yes, it's exactly what I am seeing :) Also I confirm that deleting the lib beforehand fixes it.

My local repro on linux (ArchLinux, gcc 7.2.0, GNU ar (GNU Binutils) 2.29):
* clone the repo with git
* compile everything by running `./fbuild` from the Code folder
  _all the files are isolated by default as they are all RW_
* change .UnityInputIsolateWritableFiles from true to false in Code/fbuild.bff
* compile again

```
2> /home/belkiss/RD/FastBuild/tmp/x64Linux-Release/CoreTest/CoreTest.lib(Unity1.o): In function `main':
/home/belkiss/RD/FastBuild/Code/Core/CoreTest/TestMain.cpp:12: multiple definition of `main'
```

Those steps could probably be reduced. Let me know if you want me to try and find a smaller repro :) Also if you want the version numbers of other parts of the toolchain.

PS:  (`#include <sys/uio.h>` was needed in Code/Core/Network/TCPConnectionPool.cpp so I had to remove `#if defined( __APPLE__ )`, I'll send another merge request :)

**Comment by ffulin** [comment]
I didn't manage to devise a reliable repro (your steps didn't work for me), but I did see this problem a few times. I've gone and added back deletion of the .lib file, but only if using a librarian other than lib.exe (https://github.com/fastbuild/fastbuild/commit/63121fa7226e235ea72c164902077eb46a17e5a0). This fix will be in the next release (v0.97). I'll leave this issue open until that version is released.

**Comment by ffulin** [comment]
v0.97 has now been released which contains the fix for this issue.
